fuse/file.go

```diff
@@ -3,103 +3,370 @@ package fuse
 import (
 	"context"
 	"io"
+	"os"
+	"sync"
 
 	"bazil.org/fuse"
 	"bazil.org/fuse/fs"
+	"bazil.org/fuse/fuseutil"
 	"github.com/puppetlabs/wash/activity"
 	"github.com/puppetlabs/wash/plugin"
 )
 
 // ==== FUSE file Interface ====
 
+// file's implementation differs primarily on two axes: whether `Size` is set, and whether we're
+// currently writing.
+// - a *file-like* entry declares a `Size` in its `Attributes`
+//   - reads and writes are symmetric; the kernel page cache will be used, and the file size
+//     represents the current local state of the file
+//   - when writing, `data` represents the local state of the file; changes to its size will be
+//     reflected in `data` and reads will be served from it. The file's size will be `len(data)`.
+//     Unchanged sections will be filled from `plugin.Read` as needed.
+//   - when not writing, data is read directly from `plugin.Read`
+// - a *non-file-like* entry has `Size` unset
+//   - read always pulls from `plugin.Read` and writes are buffered independently
+//   - `data` stores only the data to be written, and is not initialized from `plugin.Read`
+//   - the file's size will be reported as its readable size; it will not reflect calls to `write`
+//
+// Note that writes only result in calling `plugin.Write` when a file handle is closed. Writing
+// with multiple handles will be protected by `mux`, but all writes will operate on the same `data`
+// and the first handle close will trigger `plugin.Write`.
+//
+// `readSize` will always be initialized from either the `Size` attribute, or if unset then the
+// length of data available to read.
+//
+// `writers` are used to track in-progress writes so we know when to `plugin.Write` on `Flush`
 type file struct {
-	*fuseNode
-}
+	fuseNode
 
-var _ fs.Node = (*file)(nil)
-var _ = fs.NodeOpener(&file{})
+	mux sync.Mutex
+	// Handles with in-progress writes
+	writers map[fuse.HandleID]struct{}
+	// Only valid if len(writers) > 0
+	data []byte
+	// Size of readable content, necessary for *non-file-like* entries
+	readSize uint64
+}
 
 func newFile(p *dir, e plugin.Entry) *file {
-	return &file{newFuseNode("f", p, e)}
+	return &file{fuseNode: newFuseNode("f", p, e), writers: make(map[fuse.HandleID]struct{})}
+}
+
+func (f *file) isFileLike() bool {
```

Comment:
Hmm, maybe `isFileLike` => `isFileLikeEntry` ? It's slightly weird seeing `file#isFileLike`.

Reply:
I don't think we should restrict the definition to only things that are Writable. I think `metadata.json` specifically is file-like, because it has a consistent size.